Feature - Empty Base Layer

louwie17 opened the pull request:
This pull request adds a new feature 'emptyBaseLayer', that provides a background for the ui-grid, of continues rows, incase the height of the ui-grid is greater then the number of rows displayed.
This simulates a similar scenario as Excel, where the rows continue forever. The only exception is that the rows can't be edited, and just act as a background.
The rows/columns work with column resizing, and pinned rows.

A new tutorial page has been added. This is what it looks like in the tutorial:
![emptybaselayer](https://cloud.githubusercontent.com/assets/2240960/18710524/1366b0e6-7fdc-11e6-814e-bd06341c09a7.png)

mportuga commented on src/features/empty-base-layer/test/emptyBaseLayer.spec.js:
```diff
@@ -0,0 +1,53 @@
+
+    var $compile = _$compile_;
+    scope = $rootScope;
+    timeout = $timeout;
```
Is this timeout variable really needed? It does not get used anywhere. Even $timeout itself is only used inside this beforeEach function.

mportuga commented on src/features/empty-base-layer/test/emptyBaseLayer.spec.js:
```diff
@@ -0,0 +1,53 @@
+    $timeout(function () {
+      $compile(element)(scope);
+    });
+    $timeout.flush();
```
Why are you wrapping the compilation in a timeout? You could just as easily add a scope.$apply() function call after compiling the element instead of adding a timeout.

louwie17 replied:
Probably the only reason for doing this was going of these tests: https://github.com/angular-ui/ui-grid/blob/master/src/features/expandable/test/expandable.spec.js#L29

mportuga replied:
I will try to fix those later in that case. But in the meantime there is no reason that new code should follow the same bad practice.

louwie17 commented on src/features/empty-base-layer/js/emptyBaseLayer.js:
```diff
@@ -0,0 +1,152 @@
+         *  @description Enable empty base layer, which shows empty rows as background on the entire grid
+         *  <br/>Defaults to true, if the directive is used.
+         */
+        grid.options.enableEmptyGridBaseLayer = grid.options.enableEmptyGridBaseLayer !== false;
```
No, the two lines are different, I wanted to give the option of passing in true/false into the uiGridEmptyBaseLayer directive, like this: ui-grid-empty-base-layer="true/false". I wanted that to trump the grid.options value. That's why I check that before hand.
Unless you have a different suggestion?

mportuga replied:
I see. I can't think of another suggestion right now, but perhaps one of the other devs will.

louwie17 commented on src/features/empty-base-layer/js/emptyBaseLayer.js:
```diff
@@ -0,0 +1,152 @@
+
+      setNumberOfEmptyRows: function(viewportHeight, grid) {
+        var rowHeight = grid.options.rowHeight;
+        var rows = Math.ceil(viewportHeight / rowHeight);
```
Will make that change!

louwie17 commented on src/features/empty-base-layer/js/emptyBaseLayer.js:
```diff
@@ -0,0 +1,152 @@
+            return {
+              pre: function ($scope, $elm, $attrs, controllers) {
+              },
+              post: function ($scope, $elm, $attrs, controllers) {
```
I did this as it seemed to be the standard for the other features, mostly looked at https://github.com/angular-ui/ui-grid/blob/master/src/features/expandable/js/expandable.js#L531

mportuga replied:
Makes sense. I guess this should stay this way then.

louwie17 commented on src/features/empty-base-layer/less/emptyBaseLayer.less:
```diff
@@ -0,0 +1,9 @@
+
+.ui-grid-viewport .ui-grid-empty-base-layer-container {
+  position: absolute;
+  height: 100% !important;
```
The reason for using is because the same div inherits from 'ui-grid-canvas' as well, which gets a dynamic height/width, but I always wanted this div to take the full height of the viewport. So I should be able to find a different way.

mportuga replied:
That would be great, thanks!

mportuga commented on src/features/empty-base-layer/js/emptyBaseLayer.js:
```diff
@@ -0,0 +1,152 @@
+          emptyRows: []
+        };
+
+        grid.options.enableEmptyGridBaseLayer = baseLayerAttr !== "false";
```
undefined and null are both different than "false" couldn't that cause a problem? Is there a way for you to ensure that baseLayerAttr is a boolean when it comes in instead of a string?

louwie17 commented:
@mportuga I updated your suggested changes, see what you think? Also I squashed all the commits, hopefully that looks a little better. It was my first time doing that actually, so I had some fun screwing up the rebasing here and there :)
